src/centralstore/incremental/NodeManager.cpp

```diff
@@ -0,0 +1,254 @@
+/**
+Copyright 2020 JasmineGraph Team
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+**/
+
+#include "NodeManager.h"
+
+#include <sys/stat.h>
+
+#include "../../util/logger/Logger.h"
+#include "NodeBlock.h"  // To setup node DB
+#include "PropertyLink.h"
+#include "RelationBlock.h"
+
+Logger node_manager_logger;
+
+NodeManager::NodeManager(std::string mode) {
+    std::ios_base::openmode openMode = std::ios::trunc;  // default is Trunc mode which overrides the entire file
+    if (mode == "app") {
+        openMode = std::ios::app;  // if app, open in append mode
+        this->nodeIndex = readNodeIndex();
+    }
+    NodeBlock::nodesDB =
+        new std::fstream(NodeManager::NODE_DB_PATH, std::ios::in | std::ios::out | openMode | std::ios::binary);
+    PropertyLink::propertiesDB =
+        new std::fstream(PropertyLink::DB_PATH, std::ios::in | std::ios::out | openMode | std::ios::binary);
+    RelationBlock::relationsDB =
+        new std::fstream(RelationBlock::DB_PATH, std::ios::in | std::ios::out | openMode | std::ios::binary);
+    // TODO (tmkasun): set PropertyLink nextPropertyIndex after validating by modulus check from file number of bytes
+
+    if (dbSize(NodeManager::NODE_DB_PATH) % NodeBlock::BLOCK_SIZE != 0) {
+        node_manager_logger.error("Node DB size does not comply to node block size Path = " +
```

Comment:
We should take some action other than logging the error message here. What kind of action can the user take if this error gets thrown?

Reply:
Why don't we throw an error here like done in line 55?

Reply:
Yes agree, I think we should throw an exception here Since the program can't continue if the data is corrupted

Reply:
Do we have a JasmineGraph specific exception class ? 
I think it's good to have an exception definition for JasmineGraph.
i:e
```c++
class JasmineException: public exception
{
  virtual const char* what() const throw()
  {
    return "My exception happened";
  }
} jException;

```
~[source](http://www.cplusplus.com/doc/tutorial/exceptions/)
for the time I have used [`std::runtime_error`](https://en.cppreference.com/w/cpp/error/runtime_error) exception

Reply:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

Reply:
I think its good idea to create an Exception class for JasmineGraph. Lets use the name `JasmineGraphException`. I think we can create a top level package called exception and within that host the `JasmineGraphException` class. I will proceed creating this.

Reply:
Created [issue 117](https://github.com/miyurud/jasminegraph/issues/117) to track this. For the moment lets use the approach of throwing an error.